### PR TITLE
[Validator] Add the option filenameMaxLength to the File constraint

### LINKED ISSUE #49359: [Validator] Add an option to set a filename length limit
### Description

On this owasp [page](https://cheatsheetseries.owasp.org/cheatsheets/File_Upload_Cheat_Sheet.html) we can see `Set a filename length limit. Restrict the allowed characters if possible`.

I could be interesting to add an option in the [File Constraint](https://symfony.com/doc/current/reference/constraints/File.html) to check that ?

If yes, I can create a MR for that.

### Example

For example we can have : 
```php
// src/Entity/Author.php
namespace App\Entity;

use Symfony\Component\Validator\Constraints as Assert;

class Author
{
    #[Assert\File(
        maxSize: '1024k',
        filenameMaxLength: 50,
        filenameMaxLengthMessage: 'The filename is too long. It should have 50 character or less.'
    )]
    protected $bioFile;
}
```

### PR DESCRIPTION
**Pull request by ke20**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #49359
| License       | MIT
| Doc PR        | TODO symfony/symfony-docs#... 

This option allow to define a max length for the uploaded file:

```php
// src/Entity/Author.php
namespace App\Entity;

use Symfony\Component\Validator\Constraints as Assert;

class Author
{
    #[Assert\File(
        maxSize: '1024k',
        filenameMaxLength: 50,
        filenameMaxLengthMessage: 'The filename is too long. It should have 50 character or less.'
    )]
    protected $bioFile;
}
```

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/File.php [hunk 1]
```diff
@@ -49,13 +51,15 @@ class File extends Constraint
...
     public $mimeTypes = [];
+    public int|null $filenameMaxLength = null;
     public array|string|null $extensions = [];
     public $notFoundMessage = 'The file could not be found.';
     public $notReadableMessage = 'The file is not readable.';
     public $maxSizeMessage = 'The file is too large ({{ size }} {{ suffix }}). Allowed maximum size is {{ limit }} {{ suffix }}.';
     public $mimeTypesMessage = 'The mime type of the file is invalid ({{ type }}). Allowed mime types are {{ types }}.';
     public string $extensionsMessage = 'The extension of the file is invalid ({{ extension }}). Allowed extensions are {{ extensions }}.';
     public $disallowEmptyMessage = 'An empty file is not allowed.';
+    public string $filenameTooLongMessage = 'The filename is too long. Allowed max filename is {{ filename_max_length }} characters.';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/File.php]: I prefer the message in the description:
The filename is too long. It should have 50 character or less.

File: src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php [hunk 2]
```diff
@@ -658,5 +658,47 @@ public function testUploadedFileExtensions()
...
+
+        $this->buildViolation($messageViolation)
+            ->setParameters([
+                '{{ filename_max_length }}' => $constraintFile->filenameMaxLength,
+            ])
+            ->setCode(File::FILENAME_TOO_LONG)
+            ->assertRaised();
+    }
+
+    private function filenameMaxLengthIsTooLongProvider(): \Generator
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php]: providers should be static + this should be named provideFilenameMaxLengthIsTooLong

**Review by nicolas-grekas — COMMENTED** [review]
please have a look at test failures

File: src/Symfony/Component/Validator/Constraints/File.php (hunk 1 above)
Comment (ke20) [reply src/Symfony/Component/Validator/Constraints/File.php]: Changed ! Thank you :+1:

File: src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php (hunk 2 above)
Comment (ke20) [reply src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php]: Changed ! Thanks too !

**Comment by ke20** [comment]
> please have a look at test failures

I fixed all tests for the component `Validator`, but it seems to have some tests in failure for the others components.

**Comment by Arman-Hosseini** [comment]
@ke20 we are talking about file. So what do you think about using `nameMaxLength` instead of `filenameMaxLength`?

File: src/Symfony/Component/Validator/Constraints/File.php [hunk 3]
```diff
@@ -49,13 +51,15 @@ class File extends Constraint
 
     public $binaryFormat;
     public $mimeTypes = [];
+    public int|null $filenameMaxLength = null;
```
Comment (garak) [inline src/Symfony/Component/Validator/Constraints/File.php]: this should be `?int` instead

File: src/Symfony/Component/Validator/Constraints/File.php [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 /**
  * @Annotation
+ *
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/File.php]: I know fabbot wants you to add these lines, but that's a false positive for us :)

File: src/Symfony/Component/Validator/Constraints/File.php [hunk 5]
```diff
@@ -49,13 +52,15 @@ class File extends Constraint
 
     public $binaryFormat;
     public $mimeTypes = [];
+    public ?int $nameMaxLength = null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/File.php]: I'd keep filename everywhere - I think it makes things more obvious

File: src/Symfony/Component/Validator/Constraints/Image.php [hunk 6]
```diff
@@ -13,6 +13,7 @@
 
 /**
  * @Annotation
+ *
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/Image.php]: remove

File: src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php [hunk 7]
```diff
@@ -531,7 +531,7 @@ public function testNegativeMaxSize()
     }
 
     /**
-     * @dataProvider validExtensionProvider
+     * @dataProvider providerValidExtension
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php:534]: let's keep existing code as is

File: src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php [hunk 8]
```diff
@@ -560,7 +560,7 @@ private static function validExtensionProvider(): iterable
     }
 
     /**
-     * @dataProvider invalidExtensionProvider
+     * @dataProvider provideInvalidExtension
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/FileValidatorTestCase.php:563]: to be reverted

**Review by nicolas-grekas — APPROVED** [review]
Some nitpicking and good to go on my side.

**Comment by nicolas-grekas** [comment]
Thank you @ke20.

**Comment by ke20** [comment]
> Thank you @ke20.

Sorry I didn't have the time to fix the last little errors ... Thank you for your update !

**Pull request merged**